### PR TITLE
Display images with new image viewer (DragGesture version)

### PR DESCRIPTION
**Pull request by joshuatbrown**
## Issues covered
#1164 

## Description
Displays full-screen images with the new image viewer. The gallery view will be covered in #1177.

## How to test
1. Navigate to a post with an image
2. Tap on it
3. You're in the new image viewer!
4. Double-tap to zoom in
5. Drag around to pan
6. Double-tap again to zoom out
7. Swipe down or tap the `x` to dismiss

## Screenshots/Video
Coming soon!

## Notes
#1408 and #1413 are the same thing, but using different implementations. The point of having three draft PRs is to decide on the right approach -- each has its own pros and cons.

## Pros for this approach
1. It meets all the requirements, including double-tap to zoom and swipe to dismiss.
2. It keeps the image from being overscrolled so you don't see the edges of it like you do in #1413. (See the screenshot in the Cons section.)
3. It's pure SwiftUI.
4. No ScrollView or UIScrollView needed.

## Cons for this approach
1. Dragging feels _slightly_ sluggish as compared to the ScrollView and UIScrollView implementations.
2. Double-tapping to zoom will zoom in on the middle of the image, even if you double-tap the top left or bottom right corner. It doesn't zoom where you tapped.
3. Pinching to zoom was tricky and didn't behave as expected, so it's not implemented here. It's not a requirement right now, but if it is in the future, this approach may not work as well as the others.
4. Swiping to pan on Mac doesn't work at all. You need to click and drag.
5. The implementation of the code that keeps the image from being overscrolled is a little complex, but I could probably make it more readable.

### CODE REVIEW

File: Nos/Views/LinkPreview.swift [hunk 1]
```diff
@@ -1,47 +1,63 @@
...
-        Button(action: {
-            UIApplication.shared.open(url)
-        }, label: {
-            WebImage(url: url)
-                .resizable()
-                .indicator(.activity)
-                .aspectRatio(contentMode: .fit)
-                .frame(maxWidth: .infinity, maxHeight: .infinity)
-        })
+        Button(
```
Comment (martindsq) [inline Nos/Views/LinkPreview.swift:45]: Given that Button is a well-known component in SwiftUI, I prefer the more succinct declaration:

```
Button {
    isViewerPresented 0 true
} label: {
    WebImage()
}
```

File: Nos/Views/LinkPreview.swift [hunk 2]
```diff
@@ -1,47 +1,63 @@
...
-            
-            ZStack {
-                image
-                    .blur(radius: 10)
-                    .clipped()
-                
-                image
-                    .aspectRatio(contentMode: .fit)
-                    .frame(maxWidth: .infinity, maxHeight: .infinity)
+        Button(
```
Comment (martindsq) [inline Nos/Views/LinkPreview.swift:12]: Given that Button is a well-known component in SwiftUI, I prefer the more succinct declaration:

```
Button {
    isViewerPresented 0 true
} label: {
    ....
}
```

File: Nos/Views/LinkPreview.swift [hunk 3]
```diff
@@ -1,47 +1,63 @@
+import Dependencies
 import SwiftUI
 import LinkPresentation
 import SDWebImageSwiftUI
 
 struct ImageLinkButton: View {
     let url: URL
     
+    @State private var isViewerPresented: Bool = false
```
Comment (martindsq) [inline Nos/Views/LinkPreview.swift:9]: You can get rid of `: Bool` given that it is already implicit in `= false`

File: Nos/Views/Components/Images/ImageViewer.swift [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                    )
+                    .frame(width: geometry.size.width, height: geometry.size.height)
+                    .clipped()
+                    .ignoresSafeArea()
+            }
+
+            ZStack(alignment: .topLeading) {
+                Color.clear
+
+                Button(
```
Comment (martindsq) [inline Nos/Views/Components/Images/ImageViewer.swift:87]: Given that Button is a well-known component in SwiftUI, I prefer the more succinct declaration:

```
Button {
    dismiss()
} label: {
    ....
}
```

**Review by martindsq — CHANGES_REQUESTED** [review]

File: Nos/Views/Components/Images/ImageViewer.swift [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                    .aspectRatio(contentMode: .fit)
+                    .scaleEffect(scale)
+                    .offset(x: offset.width, y: offset.height)
+                    .gesture(
+                        DragGesture()
+                            .onChanged { value in
+                                withAnimation {
+                                    offset = CGSize(
+                                        width: lastOffset.width + value.translation.width * 1.5,
+                                        height: lastOffset.height + value.translation.height * 1.5
```
Comment (joshuatbrown) [inline Nos/Views/Components/Images/ImageViewer.swift:50]: @martindsq you said this feels weird, like the image is moving too much. We can adjust these values and see how it works for you:

```suggestion
                                        width: lastOffset.width + value.translation.width,
                                        height: lastOffset.height + value.translation.height
```

**Comment by joshuatbrown** [comment]
We're moving forward with #1413.

**Pull request closed without merging**